x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/allocation/TrainedModelAllocationNodeService.java

```diff
@@ -0,0 +1,387 @@
+        this.modelIdToTask = new ConcurrentHashMap<>();
+        this.loadingModels = new ConcurrentLinkedDeque<>();
+        this.threadPool = threadPool;
+        clusterService.addLifecycleListener(new LifecycleListener() {
+            @Override
+            public void afterStart() {
+                nodeId = clusterService.localNode().getId();
+                start();
+            }
+
+            @Override
+            public void beforeStop() {
+                stop();
+            }
+        });
+    }
+
+    TrainedModelAllocationNodeService(
+        TrainedModelAllocationService trainedModelAllocationService,
+        ClusterService clusterService,
+        DeploymentManager deploymentManager,
+        TaskManager taskManager,
+        ThreadPool threadPool,
+        String nodeId
+    ) {
+        this.trainedModelAllocationService = trainedModelAllocationService;
+        this.deploymentManager = deploymentManager;
+        this.taskManager = taskManager;
+        this.modelIdToTask = new ConcurrentHashMap<>();
+        this.loadingModels = new ConcurrentLinkedDeque<>();
+        this.threadPool = threadPool;
+        this.nodeId = nodeId;
+        clusterService.addLifecycleListener(new LifecycleListener() {
+            @Override
+            public void afterStart() {
+                start();
+            }
+
+            @Override
+            public void beforeStop() {
+                stop();
+            }
+        });
+    }
+
+    // TODO should this take a listener????
+    void stopDeployment(TrainedModelDeploymentTask task) {
+        if (stopped) {
+            return;
+        }
+        deploymentManager.stopDeployment(task);
+        taskManager.unregister(task);
+        modelIdToTask.remove(task.getModelId());
+    }
+
+    void stopDeploymentAsync(TrainedModelDeploymentTask task, ActionListener<Void> listener) {
+        threadPool.executor(MachineLearning.UTILITY_THREAD_POOL_NAME).execute(() -> {
+            try {
+                stopDeployment(task);
+                listener.onResponse(null);
+            } catch (Exception e) {
+                listener.onFailure(e);
+            }
+        });
+    }
+
+    public void start() {
+        stopped = false;
+        scheduledFuture = threadPool.scheduleWithFixedDelay(
+            this::loadQueuedModels,
+            MODEL_LOADING_CHECK_INTERVAL,
+            MachineLearning.UTILITY_THREAD_POOL_NAME
+        );
+    }
+
+    public void stop() {
+        stopped = true;
+        ThreadPool.Cancellable cancellable = this.scheduledFuture;
+        if (cancellable != null) {
+            cancellable.cancel();
+        }
+    }
+
+    void loadQueuedModels() {
+        TrainedModelDeploymentTask loadingTask;
+        logger.trace("attempting to load all currently queued models");
+        // NOTE: As soon as this method exits, the timer for the scheduler starts ticking
+        while ((loadingTask = loadingModels.poll()) != null) {
```

Comment:
We need to ensure that models are NOT loaded in parallel. Having them in a queue is nice assurance of that. But it complicates some interactions as it means that this method cannot be kicked off by the thing that detects the changes. 

Why?

Well, if it did, and we detected a second set of changes while the first `loadQueuedModels` was running we would need to not kick off the method. But, it could be that that causes the second set to not be applied at all...